# Starbucks Customer Behaviour and Promotional Campaign Analysis

## Data Mining B141 Final Project

### Name :- Sahil Pathania
### Student ID :- GH1041137

## -------------------------------------------------------

## 1. Import Libraries

In [119]:
import pandas as pd
import numpy as np 

pd.set_option('display.max.columns', None)
pd.set_option('display.max_rows', 20)

print("Libraries loaded successfully")

Libraries loaded successfully


## ---------------------------------------------------------------

## 2. Load Datasets

In [120]:
profile = pd.read_csv("../data/profile.csv")
portfolio = pd.read_csv("../data/portfolio.csv")
transcript = pd.read_csv("../data/transcript.csv")

print("Datasets loaded successfully")

Datasets loaded successfully


## -------------------------------------------------------------------------

## 3. Checking the total rows and columns in the datasets

In [121]:
print("Portfolio Shape :- ", portfolio.shape)
print("Profile Shape :- ", profile.shape)
print("transcript Shape :- ", transcript.shape)

Portfolio Shape :-  (10, 7)
Profile Shape :-  (17000, 6)
transcript Shape :-  (306534, 5)


## -------------------------------------------------------------------

## 4. Preview of Profile Dataset

In [122]:
profile.head()

,Unnamed: 0,gender,age,id,became_member_on,income
0,0,NaN,118,68be06ca386d4c31939f3a4f0e3dd783,20170212,NaN
1,1,F,55,0610b486422d4921ae7d2bf64640c50b,20170715,112000.0
2,2,NaN,118,38fe809add3b4fcf9315a9694bb96ff5,20180712,NaN
3,3,F,75,78afa995795e4d85b5d9ceeca43f5fef,20170509,100000.0
4,4,NaN,118,a03223e636434f42ac4c3df47e8bac43,20170804,NaN


## ---------------------------------------------------------------------------------------

## 5. Preview of Portfolio Dataset

In [123]:
portfolio.head()

,Unnamed: 0,reward,channels,difficulty,duration,offer_type,id
0,0,10,"['email', 'mobile', 'social']",10,7,bogo,ae264e3637204a6fb9bb56bc8210ddfd
1,1,10,"['web', 'email', 'mobile', 'social']",10,5,bogo,4d5c57ea9a6940dd891ad53e9dbe8da0
2,2,0,"['web', 'email', 'mobile']",0,4,informational,3f207df678b143eea3cee63160fa8bed
3,3,5,"['web', 'email', 'mobile']",5,7,bogo,9b98b8c7a33c4b65b9aebfe6a799e6d9
4,4,5,"['web', 'email']",20,10,discount,0b1e1539f2cc45b7b9fa7c272da2e1d7


## -------------------------------------------------------------------------------

## 6. Preview of Transcript Dataset

In [124]:
transcript.head()

,Unnamed: 0,person,event,value,time
0,0,78afa995795e4d85b5d9ceeca43f5fef,offer received,{'offer id': '9b98b8c7a33c4b65b9aebfe6a799e6d9'},0
1,1,a03223e636434f42ac4c3df47e8bac43,offer received,{'offer id': '0b1e1539f2cc45b7b9fa7c272da2e1d7'},0
2,2,e2127556f4f64592b11af22de27a7932,offer received,{'offer id': '2906b810c7d4411798c6938adc9daaa5'},0
3,3,8ec6ce2a7e7949b1bf142def7d0e0586,offer received,{'offer id': 'fafdcd668e3743c1bb461111dcafc2a4'},0
4,4,68617ca6246f4fbc85e91a2a49552598,offer received,{'offer id': '4d5c57ea9a6940dd891ad53e9dbe8da0'},0


## -------------------------------------------------------------------------------------

## 7. Structure and Data types of Dataset

In [125]:
profile.info()

<class 'pandas.DataFrame'>
RangeIndex: 17000 entries, 0 to 16999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        17000 non-null  int64  
 1   gender            14825 non-null  str    
 2   age               17000 non-null  int64  
 3   id                17000 non-null  str    
 4   became_member_on  17000 non-null  int64  
 5   income            14825 non-null  float64
dtypes: float64(1), int64(3), str(2)
memory usage: 1.3 MB


In [126]:
portfolio.info()

<class 'pandas.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype
---  ------      --------------  -----
 0   Unnamed: 0  10 non-null     int64
 1   reward      10 non-null     int64
 2   channels    10 non-null     str  
 3   difficulty  10 non-null     int64
 4   duration    10 non-null     int64
 5   offer_type  10 non-null     str  
 6   id          10 non-null     str  
dtypes: int64(4), str(3)
memory usage: 1.3 KB


In [127]:
transcript.info()

<class 'pandas.DataFrame'>
RangeIndex: 306534 entries, 0 to 306533
Data columns (total 5 columns):
 #   Column      Non-Null Count   Dtype
---  ------      --------------   -----
 0   Unnamed: 0  306534 non-null  int64
 1   person      306534 non-null  str  
 2   event       306534 non-null  str  
 3   value       306534 non-null  str  
 4   time        306534 non-null  int64
dtypes: int64(2), str(3)
memory usage: 35.0 MB


## ---------------------------------------------------------------------------------------

## 8. Statistical Summary of Numerical Variables

In [128]:
profile.describe()
print(" ")
profile[profile["age"] == 118].head()
print(" ")
profile[profile["age"] == 118].shape
print(" ")
profile[profile["age"] == 118].isnull().sum()

Unnamed: 0             0
gender              2175
age                    0
id                     0
became_member_on       0
income              2175
dtype: int64

## -------------------------------------------------------------------------------------

## 9. Analysing the missing values

In [129]:
print("Missing Values in Profile")
print(profile.isnull().sum())
print(" ")
print("Missing Values in Portfolio")
print(portfolio.isnull().sum())
print(" ")
print("Missing Values in Transcript")
print(transcript.isnull().sum())

Missing Values in Profile
Unnamed: 0             0
gender              2175
age                    0
id                     0
became_member_on       0
income              2175
dtype: int64
 
Missing Values in Portfolio
Unnamed: 0    0
reward        0
channels      0
difficulty    0
duration      0
offer_type    0
id            0
dtype: int64
 
Missing Values in Transcript
Unnamed: 0    0
person        0
event         0
value         0
time          0
dtype: int64


##  -----------------------------------------------------------------------------------

# 10. Observations from the first file "understanding"

### 1. There are 1700 customer records and 6 attributers in Profile Dataset

### 2. There are 10 promotional offers and 7 attributes in Portfolio Dataset

### 3. There are 306,534 customer activity and 5 attributes in Transcript Dataset

### 4. Missing values were identified in two columns named gender and income in Profile Dataset

### 5. The Portfolio Dataset has no missing value

### 6. Customer interactions like offer received, offer viewed, offer completed and transactions are in the transcript dataset

### 7. So during preprocessing, the column named Unnamed can be removed as it appears to be an index column 

## ----------------------------------------------------------------------------------------------

# 11. Conclusion

### 1. There are three datasets in this project names as "Profiles", "Portfolio", "Transcript".

### 2.  The profile dataset contains information like customer demographic information and includes missing values in gender and income attributes as discussed above.

### 3. The portfolio dataset consists information like promotional offer with no missing values.

### 4. The Transcript dataset has customer interactions with offers and transactions with no missing values. 

### 5. Based on my examination, this dataset fits well for further steps.